docs/rfcs/20220408_partition_hibernation.md

```diff
@@ -0,0 +1,277 @@
+it from hibernation.  On a healthy system this should be in the range
+of a few hundred milliseconds.
+
+## Alternatives
+
+### Do nothing
+
+The practical capacity of a system will be how many _active_ partitions
+we can handle: this has the advantage of determinism, as there is no
+risk of a system 
+
+### Radically improved resource management elsewhere
+
+Separately to this RFC, we know that various subsystems' resource management
+can be improved, such as the way we keep unbounded in-memory indices in
+the storage layer, and hold open file descriptors for segments.
+
+In the raft layer, hibernation would be less important if we moved to a
+model where several partitions were grouped together in one raft group,
+thereby avoiding the "heartbeat per partition" scaling issue.  This would
+be a large architectural change.
+
+### Only hibernate on explicit command from user
+
+This is viable where an external control plane has its own tracking
+of the most recent I/O on topics.  However, it requires a more
+chatty external protocol for the external control plane to learn
+the access times of partitions to figure out which ones should be
+shut down.
+
+As long as the policy for hibernating is as simple as a timeout, it
+makes sense to build it in.  We could optionally also add an API for
+hibernation so that 
+
+### Hibernate per-partition rather than per-topic
+
+There is a reasonable expectation that applications will produce
+and consume to topics rather than partitions, and per-topic hibernation
+is much simpler to expose in user interfaces, as well as writing less
+to the controller log to control it.
+
+A situation where some partitions in a topic are idle but others
+are active is a fringe case, and this feature is aimed at providing
+statistically meaningful efficiency improvements at scale: it's fine
+if we miss the opportunity to hibernate any "weird" partitions like that.
+
+### Allow reads from hibernated partitions
+
+If a partition is hibernated, its data is still present on disk, and it
+is safe to serve reads from any replica, because no writes are going
+on.
+
+If reads are going on concurrently with a partition waking up,
+this takes careful handling.  The node serving reads is not part of a raft
+group, and may not see the wakeup or the new raft group form, and thereby
+could end up serving stale reads.  To make it safe, the raft group would
+have to refuse to form until the node that had been serving reads acknowledged
+that it has noticed the partition wake up.
+
+The main reason not to do this is that reads consume system resources, so
+it contradicts the purpose of the feature.
+
+*However* we can service the special case of a fetch from an offset higher
+than the frozen offset of a the partition at time of hibernation.  That way,
+if the user leaves consumers running, they won't wake up the topic.  The topic
+can stay hibernated until a write happens.
+
+
+## Unresolved questions
+
```

Comment:
Thanks again for writing this out--you enumerate a lot of areas of complexity that this would take to implement, which is really helpful with understanding. On first read--and still learning the code--my thought is; this is a lot of complexity to add to an already complex system. Extra states, machinery, RPCs, different stats, probably customer-visible. So, my main question is, can we solve the scale problems by taking a more of a "demand allocate resources as needed to fulfill requests" approach to these various scale problems?